docs/Explore Algorithms/OpenAI/Quickstart - Custom Embeddings and Approximate KNN.ipynb

```diff
@@ -0,0 +1,550 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "42117315-a245-491a-b330-f8257d6fb35c",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "## Step 2: Load Input Data\n",
+    "\n",
+    "It will load public dataset and generate extra syntetic rows if set by size parameter\n",
+    "\n",
+    "The loaded dataset has 1000 rows. If you specify <i>number_of_input_rows</i> in [1..1000] it will cut extra rows if needed\n",
+    "\n",
+    "If <i>number_of_input_rows</i> in [1000..1000000] it will generate extra rows using cross join of original data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {
+      "byteLimit": 2048000,
+      "rowLimit": 10000
+     },
+     "inputWidgets": {},
+     "nuid": "6b6bdb2c-d492-4114-a7e9-0ef2832ac05c",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "file_path = \"wasbs://publicwasb@mmlspark.blob.core.windows.net/fine_food_reviews_1k.csv\"\n",
+    "\n",
+    "df = spark.read.options(inferSchema=\"True\", delimiter=\",\", header=True).csv(\n",
+    "    file_path\n",
+    ")\n",
+    "df = df.withColumn(\n",
+    "    \"combined\",\n",
+    "    F.format_string(\n",
+    "        \"Title: %s; Content: %s\", F.trim(df.Summary), F.trim(df.Text)\n",
+    "    ),\n",
+    ")\n",
+    "\n",
+    "# Size of DF\n",
+    "number_of_input_rows = 999\n",
+    "# Shuffle the DataFrame with a fixed seed\n",
+    "seed = 42\n",
+    "\n",
+    "# Check if the row count is less than 10\n",
+    "if number_of_input_rows <= 0 or number_of_input_rows >= 1000000:\n",
+    "    raise ValueError(f\"Limit is {number_of_input_rows}, which should be less than 1M.\")\n",
+    "\n",
+    "if number_of_input_rows > 1000:\n",
+    "\n",
+    "    # Cross-join the DataFrame with itself to create n x n pairs for string concatenation (synthetic data)\n",
+    "    cross_joined_df = df.crossJoin(\n",
+    "        df.withColumnRenamed(\"combined\", \"combined_\")\n",
+    "    )\n",
+    "\n",
+    "    # Create a new column 'result_vector' by concatenating the two source vectors\n",
+    "    tmp_df = cross_joined_df.withColumn(\n",
+    "        \"result_vector\",\n",
+    "        F.concat(F.col(\"combined\"), F.lit(\". \\n\"), F.col(\"combined_\")),\n",
+    "    )\n",
+    "\n",
+    "    # Select only the necessary columns and show the result\n",
+    "    tmp_df = tmp_df.select(\"result_vector\")\n",
+    "    df = tmp_df.withColumnRenamed(\"result_vector\", \"combined\").withColumn(\n",
+    "        \"id\", monotonically_increasing_id()\n",
+    "    )\n",
+    "\n",
+    "df = df.limit(number_of_input_rows).orderBy(rand(seed)).repartition(10).cache()\n",
+    "\n",
+    "print(f\"Loaded: {number_of_input_rows} rows\")"
```

Comment:
we probabbly can remove the cross join stuff for the demo, i would rather use a large dataset and subset than a small dataset and augment